alphadia/workflow/optimizer.py

```diff
@@ -0,0 +1,147 @@
+# native imports
+from functools import wraps
+
+import numpy as np
+
+# alpha family imports
+# third party imports
+import pandas as pd
+
+# alphadia imports
+from alphadia.workflow import reporting
+
+
+def check_if_optimum_found(method):
+    """If an optimum has been found, the optimizer should cease to execute its methods to check, initiate and update.
+    This function can be used as a decorator and placed before the methods in an optimizer to prevent execution of the methods if an optimum has been found.
+    If the optimum has been found, the decorator will print the end of optimization message and, for consistency with stop_or_continue methods, return True.
+    """
+
+    @wraps(method)
+    def wrapper(self, *args, **kwargs):
+        if self.optimal_tolerance is not None:
+            self.end_of_optimization_message()
+            return True
+        return method(self, *args, **kwargs)
+
+    return wrapper
+
+
+class BaseOptimizer:
+    def __init__(
+        self,
+        initial_tolerance: float,
+        reporter: None | reporting.Pipeline | reporting.Backend = None,
+    ):
+        """This class serves as a base class for organizing the search parameter optimization process, which defines the tolerances used for search.
+
+        Parameters
+        ----------
+        initial_tolerance: float
+            The initial tolerance to start the optimization
+        """
+        self.proposed_new_tolerance = initial_tolerance
+        self.tolerances = []
+        self.round = -1
+        self.optimal_tolerance = None
+        self.reporter = reporting.LogBackend() if reporter is None else reporter
+
+    def check_stopping_conditions(self):
+        raise NotImplementedError(
+            f"check_stopping_conditions() not implemented for {self.__class__.__name__}"
+        )
+
+    def update(self):
+        raise NotImplementedError(
+            f"update() not implemented for {self.__class__.__name__}"
+        )
+
+    def initiate(self):
+        raise NotImplementedError(f"initiate() not implemented for {self.__class__}")
+
+
+class RTOptimizer(BaseOptimizer):
+    pass
+
+
+class MS2Optimizer(BaseOptimizer):
+    def __init__(self, initial_tolerance: float):
+        super().__init__(initial_tolerance)
+        self.estimator = "mz"
```

Comment:
is this used?

Reply:
Not within the object, but I can't see how to avoid using it when setting the new parameter with calibration_manager (see https://github.com/MannLabs/alphadia/pull/283#discussion_r1691069702).